sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/Telemetry.scala

```diff
@@ -0,0 +1,147 @@
+        .setResource(resource)
+        .build()
+    val sdk = OpenTelemetrySdk
+      .builder()
+      .setTracerProvider(sdkTracerProvider)
+      .setPropagators(ContextPropagators.create(W3CTraceContextPropagator.getInstance()))
+      .build()
+    //    Runtime.getRuntime().addShutdownHook(new Thread(sdkTracerProvider.close)) TODO add shutdown
+    sdk
+  }
+
+  /**
+   * Builds the span by extracting the traceparent from the `metadata` and starts the span
+   *
+   * @param tracerName
+   * @param spanName
+   * @param metadata
+   * @return
+   */
+  def buildSpan[T](
+      component: T,
+      inputTypeUrl: String,
+      methodInvoker: Option[MethodInvoker],
+      metadata: Metadata): Option[Span] = {
+    //TODO find if we have this mapping already
+    def getComponentName[T](componentType: T): String =
+      componentType match {
+        case _: Action                   => "action"
+        case _: EventSourcedEntity[_, _] => "event-sourced-entity"
+        case _: ValueEntity[_]           => "value-entity"
+      }
+
+    def getHttpMethod(method: Method): String = {
+      val httpMethods = method.getDeclaredAnnotations.map { ann =>
+        ann match {
+          case _: org.springframework.web.bind.annotation.PostMapping   => Some(HttpMethods.POST)
+          case _: org.springframework.web.bind.annotation.GetMapping    => Some(HttpMethods.GET)
+          case _: org.springframework.web.bind.annotation.PatchMapping  => Some(HttpMethods.PATCH)
+          case _: org.springframework.web.bind.annotation.PutMapping    => Some(HttpMethods.PUT)
+          case _: org.springframework.web.bind.annotation.DeleteMapping => Some(HttpMethods.DELETE)
+          case _                                                        => None
+        }
+      }
+      httpMethods.flatten.headOption match {
+        case Some(httpMethod) => httpMethod.value
+        case None             => "No @RequestMapping annotation"
+      }
+    }
+
+    if (metadata.get(Telemetry.TRACE_PARENT_KEY).isPresent && methodInvoker.isDefined) {
+      val method = methodInvoker.get.method
+
+      val context = openTelemetry.getPropagators.getTextMapPropagator
+        .extract(OtelContext.current(), metadata, Telemetry.otelGetter.asInstanceOf[TextMapGetter[Object]])
+      val tracer = openTelemetry.getTracer("java-sdk-spring")
+      val span = tracer
+        .spanBuilder(s"""${component.getClass.getName}.${method.getName}""")
+        .setParent(context)
+        .setSpanKind(SpanKind.SERVER)
+        .startSpan()
+      Some(
+        span
+          .setAttribute(SemanticAttributes.HTTP_METHOD, getHttpMethod(method))
+          .setAttribute("server.address", "localhost") // TODO use Kubernetes .svc name?
+          .setAttribute("server.port", 8080) // TODO read the port from properties
+          .setAttribute("server.scheme", "http")
+          .setAttribute("service.name", method.getName)
```

Comment:
Is this the Java class name for the user class handling the request or is it the synthetic protobuf service method name (and is that really the "service.name")?

Reply:
changed now